File: backend/app/routers/jira.py
# Path: app/routers/jira.py
# File: jira.py
# Created: 2026-05-27
# Purpose: Read-only proxy endpoints for Jira data (projects, issues, sprints)
# Caller: app/main.py
# Callees: app/services/jira.py
# Data In: HTTP requests (no body — query-string filters only)
# Data Out: JSON (normalized issue / sprint / project dicts)
# Last Modified: 2026-05-27

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.config import settings
from app.database import get_db
from app.services import jira as svc

router = APIRouter(prefix="/api/jira", tags=["jira"])


@router.get("/config")
def get_config() -> dict:
    """Lightweight health check — does NOT contact Jira. Returns whether the
    backend has credentials configured. The frontend uses this to decide
    whether to render the Jira UI surfaces."""
    return {
        "configured": settings.jira_configured,
        "base_url": settings.JIRA_BASE_URL if settings.jira_configured else None,
        "cache_ttl_seconds": settings.JIRA_CACHE_TTL_SECONDS,
    }


@router.get("/projects")
def list_projects() -> list[dict]:
    return svc.list_projects()


@router.get("/issues/{issue_key}")
def get_issue(issue_key: str) -> dict:
    return svc.get_issue(issue_key)


@router.get("/search")
def search_issues(
    jql: str = Query(..., description="Raw JQL query"),
    limit: int = Query(50, ge=1, le=100),
) -> list[dict]:
    return svc.search_issues(jql=jql, limit=limit)


@router.get("/projects/{project_key}/sprints")
def list_active_sprints(project_key: str) -> list[dict]:
    return svc.get_active_sprints(project_key)


@router.get("/sprints/{sprint_id}/issues")
def list_sprint_issues(sprint_id: int) -> list[dict]:
    return svc.get_sprint_issues(sprint_id)


@router.post("/cache/clear", status_code=204)
def clear_cache() -> None:
    """Admin escape hatch — drop all cached Jira responses."""
    svc.clear_cache()


@router.post("/sync")
def sync_from_jira(
    project_id: int | None = Query(None, description="Restrict to one project; omit for all"),
    db: Session = Depends(get_db),
) -> dict:
    """Mirror Jira issue status into linked DWB tickets. Jira leads.

    Returns a summary: counts of synced/changed/unmapped/errors and per-row
    details for the changed tickets.
    """
    return svc.sync_linked_tickets(db, project_id=project_id)


@router.get("/rollup")
def jira_rollup(
    project_id: int = Query(..., description="DWB project id"),
    db: Session = Depends(get_db),
) -> dict:
    """Group a project's linked tickets by their root Jira Epic and return
    per-epic completion stats. Walks one hop up for Subtask parents."""
    return svc.rollup_by_epic(db, project_id=project_id)

File: frontend/src/api/jira.js
// Path: src/api/jira.js
// File: jira.js
// Created: 2026-05-27
// Purpose: API wrappers for the /api/jira/* live proxy endpoints (read-only Jira data) plus DWB-342's three project-scoped snapshot endpoints (cached Jira table + manual sync trigger + sync-status poll)
// Caller: pages/JiraIssuesPage.jsx, components/tickets/TicketList.jsx (badge), store/useStore.js (jira slice)
// Callees: ./client (get, post)
// Data In: Issue key, JQL string, project key, sprint id, project_id for snapshot endpoints
// Data Out: Normalized issue / sprint / project dicts; snapshot row arrays

import { get, post } from './client';

export function getJiraConfig() {
  return get('/jira/config');
}

export function listJiraProjects() {
  return get('/jira/projects');
}

export function getJiraIssue(issueKey) {
  return get(`/jira/issues/${issueKey}`);
}

export function searchJiraIssues(jql, limit = 50) {
  return get('/jira/search', { jql, limit });
}

export function listActiveSprints(projectKey) {
  return get(`/jira/projects/${projectKey}/sprints`);
}

export function listSprintIssues(sprintId) {
  return get(`/jira/sprints/${sprintId}/issues`);
}

export function clearJiraCache() {
  return post('/jira/cache/clear');
}

export function syncFromJira(projectId = null) {
  const qs = projectId != null ? `?project_id=${projectId}` : '';
  return post(`/jira/sync${qs}`);
}

// getJiraRollup removed in DWB-342 — JiraRollupPage was deleted as part of
// the unification (single Jira nav link, single table). Backend /jira/rollup
// endpoint may still exist; no frontend caller.

// ---- DWB-342: snapshot-backed project Jira table ----

export function getProjectJiraTickets(projectId, params = {}) {
  return get(`/projects/${projectId}/jira-tickets`, params);
}

export function triggerProjectJiraSync(projectId) {
  return post(`/projects/${projectId}/jira-sync`);
}

export function getProjectJiraSyncStatus(projectId) {
  return get(`/projects/${projectId}/jira-sync/status`);
}
